src/modules/rest/impl/src/main/java/it/geosolutions/geostore/services/rest/security/GeoStoreLdapAuthoritiesPopulator.java

```diff
@@ -184,18 +216,32 @@ public void setAuthoritiesMapper(GrantedAuthoritiesMapper authoritiesMapper) {
         return getGroupsOrRoles(null, null, false, true);
     }
     
+    private void addAuthorities(String[] params, Set<GrantedAuthority> authorities,
+            String root, String filter, String authorityPrefix, boolean hierarchical) {
+        addAuthorities(params, authorities, root, filter, authorityPrefix, hierarchical, 0);
+    }
     
     private void addAuthorities(String[] params, Set<GrantedAuthority> authorities,
-            String root, String filter, String authorityPrefix) {
-        Set<String> ldapAuthorities = ldapTemplate.searchForSingleAttributeValues(root, filter,
-                params, groupRoleAttribute);
+            String root, String filter, String authorityPrefix, boolean hierarchical, int level) {
+        String formattedFilter = MessageFormat.format(filter, params);
         
+        List ldapAuthorities = ldapTemplate.search(root, formattedFilter, new AbstractContextMapper() {
+            @Override
+            protected Object doMapFromContext(DirContextOperations ctx) {
+                return new Authority(ctx.getStringAttribute(groupRoleAttribute), ctx.getNameInNamespace());
+            }
+        });
+
         if (logger.isDebugEnabled()) {
             logger.debug("Authorities from search: " + ldapAuthorities);
         }
-        for (String authority : ldapAuthorities) {
-
-            addAuthority(authorities, authorityPrefix, authority);
+        for (Object authority : ldapAuthorities) {
+        	Authority ldapAuthority = (Authority)authority;
+            addAuthority(authorities, authorityPrefix, ldapAuthority.getName());
+            if (hierarchical && level < maxLevelGroupsSearch) {
```

Comment:
As probable improvement, here could we check also if current ldap group isn't already on Set<GrantedAuthority> authorities, for avoiding possible Ldap (bad?) circular references and long recursive calls?

Asking this since I'm checking it on GS side.

Reply:
Yes, that's absolutely possible. I will add this additional check. thanks

Reply:
I added the check